Lib/imaplib.py

```diff
@@ -588,6 +648,48 @@ def getquotaroot(self, mailbox):
         return typ, [quotaroot, quota]
 
 
+    def idle(self, duration=None):
+        """Return an Idler: an iterable context manager for the IDLE command
+
+        :param duration:  Maximum duration (in seconds) to keep idling,
+                          or None for no time limit.
+                          To avoid inactivity timeouts on servers that impose
+                          them, callers are advised to keep this <= 29 minutes.
+                          See the note below regarding IMAP4_stream on Windows.
+        :type duration:   int|float|None
+
+        The context manager sends the IDLE command upon entry, produces
+        responses via iteration, and sends DONE upon exit.
+        It represents responses as (type, datum) tuples, rather than the
+        (type, [data, ...]) tuples returned by other methods, because only one
+        response is represented at a time.
+
+        Example:
+
+        with imap.idle(duration=29 * 60) as idler:
+            for typ, datum in idler:
+                print(typ, datum)
+
+        Responses produced by the iterator are not added to the internal
+        cache for retrieval by response().
+
+        Note: Windows IMAP4_stream connections have no way to accurately
+        respect 'duration', since Windows select() only works on sockets.
+        However, if the server regularly sends status messages during IDLE,
+        they will wake our selector and keep iteration from blocking for long.
+        Dovecot's imap_idle_notify_interval is two minutes by default.
+        Assuming that's typical of IMAP servers, subtracting it from the 29
+        minutes needed to avoid server inactivity timeouts would make 27
+        minutes a sensible value for 'duration' in this situation.
+
+        Note: The Idler class name and structure are internal interfaces,
+        subject to change.  Calling code can rely on its context management,
+        iteration, and public method to remain stable, but should not
+        subclass, instantiate, or otherwise directly reference the class.
```

Comment:
Would checking `isinstance(x, imaplib.Idler)` be acceptable use? In that case you should probably commit to keeping the class name stable.